# Chapter 5 - Bit Manipulation

## Most significant bit (MSB)

## Bit to int conversions

In [6]:
def msb(num, neg=True):
    sum = 0
    for i in range(0, 65):
        if sum >= num: 
            if neg: return i + 1
            return i
        sum += 2**i

num = 13
print(msb(num))


5


In [45]:
def int2bin(val, bits = None, curr='', pos=0, sep=''):
    if bits == None: bits = msb(val)
    if pos == bits: return curr

    exp = bits-pos-1
    num = 2**exp

    if val < 0:
        val = 2**exp + val
        curr += '1'+sep
    elif pos == 0 and val >= 0:
        val = val
        curr += '0'+sep
    elif val >= num:
        val = val - num
        curr += '1'
    else:
        curr += '0'
    result = int2bin(val, bits, curr, pos+1)

    #for i in range(0, bits - sig):
        #result = result[0:i] + '0' + result[i+1:len(result)]
    return result

n, b = -7, 4
print(n, int2bin(n,b))
n, b = -1, 4
print(n, int2bin(n,b))
n, b = 0, 4
print(n, int2bin(n,b))
n, b = 5, 4
print(n, int2bin(n,b))
n, b = 19, 10
print(n, int2bin(n,b))
n, b = -125, 12
print(n, int2bin(n,b))

-7 1001
-1 1111
0 0000
5 0101
19 0000010011
-125 111110000011


In [ ]:
def int2bin2(num, bits = None, curr='', pos=0, sep=''):
    


In [46]:
def bin2int(binary, hassign = True):
    bits = len(binary)
    result = 0
    maxbit = 2**(bits-1) 
    if hassign and binary[0] == '1': 
        val = bin2int(binary[1:], False)
        result = val - maxbit
        return result
    elif hassign:
        binary = binary[1:]

    for i in range(len(binary)-1, -1, -1):
        j = len(binary) - 1 - i
        b = int(binary[i])
        result += b * 2**j

    return result

b = '0101'
print(b, bin2int(b))
b = '1111'
print(b, bin2int(b))
b = '1101'
print(b, bin2int(b))
b = '11011010'
res = bin2int(b)
print(b, bin2int(b))
print(res, int2bin(res,len(b)))
b = '00010011'
print(b, bin2int(b))
b = '111110000011'
print(b, bin2int(b))

0101 5
1111 -1
1101 -3
11011010 -38
-38 11011010
00010011 19
111110000011 -125


## Common bit tasks

In [9]:
# getbit: bit at i from right == 1 (boolean)
def getbit(num, i, verbose = False, bits=None):

    if verbose: 
        if bits == None: bits = msb(num)
        print('getbit')
        print('\tnum=\t', int2bin(num, bits), num)
        print('\t1 << i=\t', int2bin(1 << i, bits), 1 <<i)
        print('num & (1 << i)=\t', int2bin(num & (1 << i),bits),num & (1 << i))
        print()
    
    return ((num & (1 << i)) != 0)

b = '0101'
num = bin2int(b)
i = 0
print('bin=' + b,'num='+str(num), 'i='+str(i), 'getbit=' + str(getbit(num,i, 4, False)))
i = 1
print('bin=' + b,'num='+str(num), 'i='+str(i), 'getbit=' + str(getbit(num,i, 4)))
i = 2
print('bin=' + b,'num='+str(num), 'i='+str(i), 'getbit=' + str(getbit(num,i, 4)))

getbit
	num=	  5
	1 << i=	  1
num & (1 << i)=	  1

bin=0101 num=5 i=0 getbit=True
getbit
	num=	 0101 5
	1 << i=	 0010 2
num & (1 << i)=	 0000 0

bin=0101 num=5 i=1 getbit=False
getbit
	num=	 0101 5
	1 << i=	 0100 4
num & (1 << i)=	 0100 4

bin=0101 num=5 i=2 getbit=True


In [10]:
# setbit: set bit at i from right == 1 (new int)
def setbit(num, i, verbose=False,  bits=8,):
    if verbose: 
        if bits == None: bits = msb(num)
        print('setbit')
        print('\tnum=\t', int2bin(num, bits), num)
        print('\t1 << i=\t', int2bin(1 << i, bits), 1 <<i)
        print('num | (1 << i)=\t', int2bin(num | (1 << i),bits),num | (1 << i))
        print()
    return num | (1 << i)

b = '0101'
num = bin2int(b)
i = 1
print('bin=' + b,'num='+str(num), 'i='+str(i), 'getbit=' + str(setbit(num,i, 4, True)))

setbit
	num=	 0 5
	1 << i=	 0 2
num | (1 << i)=	 0 7

bin=0101 num=5 i=1 getbit=7


In [40]:
# clear bit: set bit at i from right == 0 (new int)
def clearbit(num, i):
    mask = ~(1 << i)
    return num & mask

def clearmsbthrui(num, i, verbose=False, bits = None):
    if bits == None: bits == msb(num)
    if verbose: print('num=',int2bin(num, bits))
    mask = (1 << i) - 1
    if verbose: print('mask=',int2bin(mask, bits))
    result = num & mask
    if verbose: print('result=',int2bin(result, bits),'\n')
    return result

clearmsbthrui(5,2, True, 4)

def clearithru0(num, i, verbose=False, bits=None):
    if bits == None: bits = msb(num)
    if verbose: print('num=', int2bin(num,bits))
    mask = (-1 << (i+1))
    if verbose: print('mask=',int2bin(mask,bits))
    result = num & mask
    if verbose: print('result=',int2bin(result,bits),'\n')
    return result

clearithru0(num, 1, True, 4)

num= 0101
mask= 0011
result= 0001 

num= 0101
mask= 1100
result= 0100 



4

In [47]:
def updatebit(num, int, bitval=1, verbose=True, bits=None):
    if bits == None: bits = msb(num)
    if verbose: print('num=', int2bin(num,bits))
    mask = ~(1 << i)
    if verbose: print('mask=',int2bin(mask,bits))
    nummask = num & mask
    if verbose: print('nummask=',int2bin(nummask,bits))
    vallsi = bitval << i
    if verbose: print('vallsi=',int2bin(vallsi,bits))
    result = nummask | vallsi
    if verbose: print('result=',int2bin(result,bits),'\n')

    return result

updatebit(5,1,1,True)

num= 0101
mask= 1011
nummask= 0001
vallsi= 0100
result= 0101 



5

## Questions

### 5.1 Insertion: 

You are given two 32-bit numbers, Nand M, and two bit positions, i and
j. Write a method to insert Minto N such that M starts at bit j and ends at bit i. You
can assume that the bits j through i have enough space to fit all of M. That is, if
M = 10011, you can assume that there are at least 5 bits between j and i. You would not, for
example, have j = 3 and i = 2, because M could not fully fit between bit 3 and bit 2.


EXAMPLE

Input: N = 10000000000, M 10011, i 2, j 6

Output: N 10001001100

In [51]:
def insertion(n, m, i, j, bits = None):
    if bits == None: bits = msb(n)
    print('m=', int2bin(m, bits))
    allones = ~0
    print('allones=', int2bin(allones, bits))
    left = allones << (j+1)
    print('left=', int2bin(left, bits))
    right = (1 << i) -1 
    print('right=', int2bin(right, bits))
    mask = left | right
    print('mask=', int2bin(mask, bits), bin(mask))
    nclear = n & mask
    print('nclear=', int2bin(nclear, bits))
    mshift = m << i
    print('mshift=', int2bin(mshift, bits))
    result = nclear | mshift
    print('result=', int2bin(result, bits))



nbin = '010000011001'
mbin = '010011'
n = bin2int(nbin)
m = bin2int(mbin)
bits = len(nbin)
i = 2
j = 6

insertion(n, m, i, j, bits)

m= 000000010011
allones= 111111111111
left= 111110000000
right= 000000000011
mask= 111110000011 -0b1111101
nclear= 010000000001
mshift= 000001001100
result= 010001001101
